src/main/java/se/sundsvall/webmessagecollector/service/scheduler/MessageCacheService.java

```diff
@@ -40,41 +42,78 @@ public class MessageCacheService {
 	}
 
 	@Transactional
-	public List<MessageEntity> fetchMessages(final String municipalityId, final Instance instance, final String familyId) {
+	public List<MessageEntity> fetchAndSaveMessages(final String municipalityId, final Instance instance, final String familyId, final Duration clockSkew) {
 		// Fetch info regarding last execution for fetching familyId (or initiate entity if no info exists)
 		var executionInfo = executionInformationRepository.findById(familyId).orElse(initiateExecutionInfo(municipalityId, familyId));
 		// Calculate timestamp from when messages should be fetched
-		var fromTimestamp = executionInfo.getLastSuccessfulExecution().format(DATE_TIME_FORMAT);
+		var fromTimestamp = executionInfo.getLastSuccessfulExecution().minus(clockSkew).format(DATE_TIME_FORMAT);
 
+		var startTime = OffsetDateTime.now();
 		var bytes = openEIntegration.getMessages(municipalityId, instance, familyId, fromTimestamp, "");
-		var messages = toMessageEntities(municipalityId, bytes, familyId, instance);
+		var messages = toMessageEntities(municipalityId, bytes, familyId, instance).stream()
+			.filter(this::notFetchedPreviously)
+			.toList();
 		messageRepository.saveAllAndFlush(messages);
 
 		// Update timestamp for last successful execution
-		executionInfo.setLastSuccessfulExecution(OffsetDateTime.now());
+		executionInfo.setLastSuccessfulExecution(startTime);
 
 		executionInformationRepository.save(executionInfo);
 		return messages;
 	}
 
 	@Transactional
-	public void fetchAttachment(final String municipalityId, final Instance instance, final MessageAttachmentEntity attachmentEntity) {
+	public void fetchAndSaveAttachment(final String municipalityId, final Instance instance, final MessageAttachmentEntity attachmentEntity) {
 		try {
-			var attachmentStream = openEIntegration.getAttachment(municipalityId, instance, attachmentEntity.getAttachmentId());
-			if (attachmentStream != null) {
-				attachmentEntity.setFile(new SerialBlob(attachmentStream));
+			var attachmentByteArray = openEIntegration.getAttachment(municipalityId, instance, attachmentEntity.getAttachmentId());
+			if (attachmentByteArray != null && attachmentEntity.getFile() == null) {
+				attachmentEntity.setFile(new SerialBlob(attachmentByteArray));
 				messageAttachmentRepository.saveAndFlush(attachmentEntity);
 			}
-		} catch (SQLException e) {
-			LOG.error("Unable to fetch attachment ", e);
+		} catch (Exception e) {
+			LOG.error("Unable to fetch attachment: municipalityId:'{}', instance:'{}', entity:'{}' ", municipalityId, instance, attachmentEntity);
+			throw new RuntimeException(e);
```

Comment:
Throw something cooler than this. 😄

Reply:
Wrapping in RuntimeException since caller uses stream and cannot handle checked exceptions. What would be cooler?

Reply:
Fair enough.